File: application/src/main/java/com/example/server/mapper/CommentMapper.java
package com.example.server.mapper;

import com.example.api.dto.CommentDto;
import com.example.server.repository.entity.Comment;
import com.example.server.repository.entity.Item;
import com.example.server.repository.entity.User;
import org.springframework.stereotype.Component;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Component
public class CommentMapper {

    public Comment toComment(CommentDto commentDto, Item item, User author) {
        if ( commentDto == null && item == null && author == null ) {
            return null;
        }

        Comment comment = new Comment();

        if ( commentDto != null ) {
            comment.setText( commentDto.getText() );
        }
        comment.setItem( item );
        comment.setAuthor( author );

        return comment;
    }

    public CommentDto toCommentDto(Comment comment) {
        if ( comment == null ) {
            return null;
        }

        String authorName = null;
        Long id = null;
        String text = null;
        LocalDateTime created = null;

        authorName = commentAuthorName( comment );
        id = comment.getId();
        text = comment.getText();
        created = comment.getCreated();

        CommentDto commentDto = new CommentDto( id, text, authorName, created );

        return commentDto;
    }

    public List<CommentDto> toCommentDto(List<Comment> comments) {
        if ( comments == null ) {
            return null;
        }

        List<CommentDto> list = new ArrayList<CommentDto>( comments.size() );
        for ( Comment comment : comments ) {
            list.add( toCommentDto( comment ) );
        }

        return list;
    }

    private String commentAuthorName(Comment comment) {
        User author = comment.getAuthor();
        if ( author == null ) {
            return null;
        }
        return author.getName();
    }
}

File: application/src/main/kotlin/com/example/server/mapper/CommentMapper.kt
package com.example.server.mapper

import com.example.api.model.CommentDto
import com.example.server.entity.Comment
import com.example.server.entity.Item
import com.example.server.entity.User

fun CommentDto.toEntity(item: Item, author: User): Comment {
    return Comment(id = this.id, text = this.text, item = item, author = author, created = null)
}

fun Comment.toDto(): CommentDto {
    return CommentDto(id = this.id, text = this.text, authorName = this.author.name, created = this.created)
}

fun List<Comment>.toDto(): List<CommentDto> {
    return map { it.toDto() }
}
